aws/resource_aws_iot_topic_rule.go

```diff
@@ -0,0 +1,558 @@
+				RoleArn:            aws.String(raw["role_arn"].(string)),
+			},
+		}
+		i++
+	}
+
+	// Add Kinesis actions
+
+	for _, a := range kinesisActions {
+		raw := a.(map[string]interface{})
+		actions[i] = &iot.Action{
+			Kinesis: &iot.KinesisAction{
+				RoleArn:      aws.String(raw["role_arn"].(string)),
+				StreamName:   aws.String(raw["stream_name"].(string)),
+				PartitionKey: aws.String(raw["partition_key"].(string)),
+			},
+		}
+		i++
+	}
+
+	// Add Lambda actions
+
+	for _, a := range lambdaActions {
+		raw := a.(map[string]interface{})
+		actions[i] = &iot.Action{
+			Lambda: &iot.LambdaAction{
+				FunctionArn: aws.String(raw["function_arn"].(string)),
+			},
+		}
+		i++
+	}
+
+	// Add Republish actions
+
+	for _, a := range republishActions {
+		raw := a.(map[string]interface{})
+		actions[i] = &iot.Action{
+			Republish: &iot.RepublishAction{
+				RoleArn: aws.String(raw["role_arn"].(string)),
+				Topic:   aws.String(raw["topic"].(string)),
+			},
+		}
+		i++
+	}
+
+	// Add S3 actions
+
+	for _, a := range s3Actions {
+		raw := a.(map[string]interface{})
+		actions[i] = &iot.Action{
+			S3: &iot.S3Action{
+				BucketName: aws.String(raw["bucket_name"].(string)),
+				Key:        aws.String(raw["key"].(string)),
+				RoleArn:    aws.String(raw["role_arn"].(string)),
+			},
+		}
+		i++
+	}
+
+	// Add SNS actions
+
+	for _, a := range snsActions {
+		raw := a.(map[string]interface{})
+		actions[i] = &iot.Action{
+			Sns: &iot.SnsAction{
+				RoleArn:       aws.String(raw["role_arn"].(string)),
+				TargetArn:     aws.String(raw["target_arn"].(string)),
+				MessageFormat: aws.String(raw["message_format"].(string)),
+			},
+		}
+		i++
+	}
+
+	// Add SQS actions
+
+	for _, a := range sqsActions {
+		raw := a.(map[string]interface{})
+		actions[i] = &iot.Action{
+			Sqs: &iot.SqsAction{
+				QueueUrl:  aws.String(raw["queue_url"].(string)),
+				RoleArn:   aws.String(raw["role_arn"].(string)),
+				UseBase64: aws.Bool(raw["use_base64"].(bool)),
+			},
+		}
+		i++
+	}
+
+	return &iot.TopicRulePayload{
+		Description:      aws.String(d.Get("description").(string)),
+		RuleDisabled:     aws.Bool(!d.Get("enabled").(bool)),
+		Sql:              aws.String(d.Get("sql").(string)),
+		AwsIotSqlVersion: aws.String(d.Get("sql_version").(string)),
+		Actions:          actions,
+	}
+}
+
+func resourceAwsIotTopicRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	ruleName := d.Get("name").(string)
+
+	_, err := conn.CreateTopicRule(&iot.CreateTopicRuleInput{
+		RuleName:         aws.String(ruleName),
+		TopicRulePayload: createTopicRulePayload(d),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId(ruleName)
+
+	return resourceAwsIotTopicRuleRead(d, meta)
+}
+
+func resourceAwsIotTopicRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).iotconn
+
+	out, err := conn.GetTopicRule(&iot.GetTopicRuleInput{
```

Comment:
> Also, we should check that the rule actually exists

Do you mean that we should do some sort of pre-flight check for the rule? The code right now faithfully looks for error returns from `GetTopicRule`, so that should cover the case where the rule doesn't exist.